### PR TITLE
Fix libuv core portability tests

### LINKED ISSUE #10879: Many c_linux_dbg_uv tests are failing (in Linux Portability Master)
Linux portability PR only checks if the tests build correctly (--build_only), but on master, many of these tests have been failing.

```
22:32:13,250 FAILED: bins/dbg/h2_census_test ping_pong_streaming
2017-04-26 22:32:13,250 FAILED: bins/dbg/h2_compress_nosec_test hpack_size
2017-04-26 22:32:13,250 FAILED: bins/dbg/h2_compress_nosec_test large_metadata
2017-04-26 22:32:13,250 FAILED: bins/dbg/h2_compress_nosec_test max_connection_age
2017-04-26 22:32:13,250 FAILED: bins/dbg/h2_compress_nosec_test simple_request
2017-04-26 22:32:13,250 FAILED: bins/dbg/h2_compress_nosec_test streaming_error_response
2017-04-26 22:32:13,251 FAILED: bins/dbg/h2_compress_test cancel_after_accept
2017-04-26 22:32:13,251 FAILED: bins/dbg/h2_compress_test compressed_payload
2017-04-26 22:32:13,251 FAILED: bins/dbg/h2_compress_test hpack_size
2017-04-26 22:32:13,251 FAILED: bins/dbg/h2_compress_test idempotent_request
2017-04-26 22:32:13,251 FAILED: bins/dbg/h2_compress_test invoke_large_request
2017-04-26 22:32:13,251 FAILED: bins/dbg/h2_compress_test ping_pong_streaming
2017-04-26 22:32:13,251 FAILED: bins/dbg/h2_compress_test registered_call
2017-04-26 22:32:13,251 FAILED: bins/dbg/h2_compress_test simple_request
2017-04-26 22:32:13,251 FAILED: bins/dbg/h2_compress_test streaming_error_response
2017-04-26 22:32:13,251 FAILED: bins/dbg/h2_fakesec_test hpack_size
2017-04-26 22:32:13,251 FAILED: bins/dbg/h2_fakesec_test idempotent_request
2017-04-26 22:32:13,251 FAILED: bins/dbg/h2_fakesec_test load_reporting_hook
2017-04-26 22:32:13,251 FAILED: bins/dbg/h2_fakesec_test max_concurrent_streams
2017-04-26 22:32:13,251 FAILED: bins/dbg/h2_fakesec_test max_connection_age
2017-04-26 22:32:13,251 FAILED: bins/dbg/h2_fakesec_test max_message_length
2017-04-26 22:32:13,251 FAILED: bins/dbg/h2_fakesec_test no_logging
2017-04-26 22:32:13,251 FAILED: bins/dbg/h2_fakesec_test ping
2017-04-26 22:32:13,251 FAILED: bins/dbg/h2_fakesec_test ping_pong_streaming
2017-04-26 22:32:13,252 FAILED: bins/dbg/h2_fakesec_test registered_call
2017-04-26 22:32:13,252 FAILED: bins/dbg/h2_fakesec_test request_with_payload
2017-04-26 22:32:13,252 FAILED: bins/dbg/h2_fakesec_test server_finishes_request
2017-04-26 22:32:13,252 FAILED: bins/dbg/h2_fakesec_test shutdown_finishes_calls
2017-04-26 22:32:13,252 FAILED: bins/dbg/h2_fakesec_test simple_cacheable_request
2017-04-26 22:32:13,252 FAILED: bins/dbg/h2_fakesec_test simple_request
2017-04-26 22:32:13,252 FAILED: bins/dbg/h2_full+trace_nosec_test binary_metadata
2017-04-26 22:32:13,252 FAILED: bins/dbg/h2_full+trace_nosec_test cancel_after_accept
2017-04-26 22:32:13,252 FAILED: bins/dbg/h2_full+trace_nosec_test cancel_after_client_done
2017-04-26 22:32:13,252 FAILED: bins/dbg/h2_full+trace_nosec_test compressed_payload
2017-04-26 22:32:13,252 FAILED: bins/dbg/h2_full+trace_nosec_test filter_call_init_fails
2017-04-26 22:32:13,252 FAILED: bins/dbg/h2_full+trace_nosec_test filter_latency
2017-04-26 22:32:13,252 FAILED: bins/dbg/h2_full+trace_nosec_test graceful_server_shutdown
2017-04-26 22:32:13,252 FAILED: bins/dbg/h2_full+trace_nosec_test high_initial_seqno
2017-04-26 22:32:13,252 FAILED: bins/dbg/h2_full+trace_nosec_test idempotent_request
2017-04-26 22:32:13,252 FAILED: bins/dbg/h2_full+trace_nosec_test keepalive_timeout
2017-04-26 22:32:13,252 FAILED: bins/dbg/h2_full+trace_nosec_test load_reporting_hook
2017-04-26 22:32:13,252 FAILED: bins/dbg/h2_full+trace_nosec_test max_connection_age
2017-04-26 22:32:13,253 FAILED: bins/dbg/h2_full+trace_nosec_test max_message_length
2017-04-26 22:32:13,253 FAILED: bins/dbg/h2_full+trace_nosec_test ping
2017-04-26 22:32:13,253 FAILED: bins/dbg/h2_full+trace_nosec_test ping_pong_streaming
2017-04-26 22:32:13,253 FAILED: bins/dbg/h2_full+trace_nosec_test registered_call
2017-04-26 22:32:13,253 FAILED: bins/dbg/h2_full+trace_nosec_test server_finishes_request
2017-04-26 22:32:13,253 FAILED: bins/dbg/h2_full+trace_nosec_test shutdown_finishes_calls
2017-04-26 22:32:13,253 FAILED: bins/dbg/h2_full+trace_nosec_test simple_cacheable_request
2017-04-26 22:32:13,253 FAILED: bins/dbg/h2_full+trace_nosec_test simple_delayed_request
2017-04-26 22:32:13,253 FAILED: bins/dbg/h2_full+trace_nosec_test simple_request
2017-04-26 22:32:13,253 FAILED: bins/dbg/h2_full+trace_nosec_test streaming_error_response
2017-04-26 22:32:13,253 FAILED: bins/dbg/h2_full+trace_nosec_test write_buffering
2017-04-26 22:32:13,253 FAILED: bins/dbg/h2_full+trace_test filter_call_init_fails
2017-04-26 22:32:13,253 FAILED: bins/dbg/h2_full+trace_test invoke_large_request
2017-04-26 22:32:13,253 FAILED: bins/dbg/h2_full+trace_test registered_call
2017-04-26 22:32:13,253 FAILED: bins/dbg/h2_full+trace_test simple_request
2017-04-26 22:32:13,253 FAILED: bins/dbg/h2_full+trace_test streaming_error_response
2017-04-26 22:32:13,253 FAILED: bins/dbg/h2_full_nosec_test hpack_size
2017-04-26 22:32:13,253 FAILED: bins/dbg/h2_full_nosec_test idempotent_request
2017-04-26 22:32:13,254 FAILED: bins/dbg/h2_full_nosec_test invoke_large_request
2017-04-26 22:32:13,254 FAILED: bins/dbg/h2_full_nosec_test keepalive_timeout
2017-04-26 22:32:13,254 FAILED: bins/dbg/h2_full_nosec_test no_logging
2017-04-26 22:32:13,254 FAILED: bins/dbg/h2_full_nosec_test ping_pong_streaming
2017-04-26 22:32:13,254 FAILED: bins/dbg/h2_full_nosec_test request_with_payload
2017-04-26 22:32:13,254 FAILED: bins/dbg/h2_full_nosec_test simple_request
2017-04-26 22:32:13,254 FAILED: bins/dbg/h2_full_nosec_test trailing_metadata
2017-04-26 22:32:13,254 FAILED: bins/dbg/h2_full_test bad_ping
2017-04-26 22:32:13,254 FAILED: bins/dbg/h2_full_test binary_metadata
2017-04-26 22:32:13,254 FAILED: bins/dbg/h2_full_test cancel_after_accept
2017-04-26 22:32:13,254 FAILED: bins/dbg/h2_full_test idempotent_request
2017-04-26 22:32:13,254 FAILED: bins/dbg/h2_full_test keepalive_timeout
2017-04-26 22:32:13,254 FAILED: bins/dbg/h2_full_test large_metadata
2017-04-26 22:32:13,254 FAILED: bins/dbg/h2_full_test max_message_length
2017-04-26 22:32:13,254 FAILED: bins/dbg/h2_full_test ping
2017-04-26 22:32:13,254 FAILED: bins/dbg/h2_full_test shutdown_finishes_calls
2017-04-26 22:32:13,254 FAILED: bins/dbg/h2_full_test simple_request
2017-04-26 22:32:13,254 FAILED: bins/dbg/h2_full_test streaming_error_response
2017-04-26 22:32:13,255 FAILED: bins/dbg/h2_load_reporting_nosec_test filter_latency
2017-04-26 22:32:13,255 FAILED: bins/dbg/h2_load_reporting_nosec_test hpack_size
2017-04-26 22:32:13,255 FAILED: bins/dbg/h2_load_reporting_nosec_test ping_pong_streaming
2017-04-26 22:32:13,255 FAILED: bins/dbg/h2_ssl_cert_test cancel_after_client_done
2017-04-26 22:32:13,255 FAILED: bins/dbg/h2_ssl_cert_test cancel_in_a_vacuum
2017-04-26 22:32:13,255 FAILED: bins/dbg/h2_ssl_cert_test empty_batch
2017-04-26 22:32:13,255 FAILED: bins/dbg/h2_ssl_cert_test registered_call
2017-04-26 22:32:13,255 FAILED: bins/dbg/h2_ssl_cert_test simple_cacheable_request
2017-04-26 22:32:13,255 FAILED: bins/dbg/h2_ssl_test call_creds
2017-04-26 22:32:13,543 START: tools/run_tests/helper_scripts/post_tests_c.sh
2017-04-26 22:32:13,548 PASSED: tools/run_tests/helper_scripts/post_tests_c.sh [time=0.0sec; retries=0:0]
2017-04-26 22:32:13,551 FAILED: Some tests failed
```

https://grpc-testing.appspot.com/job/gRPC_portability_master_linux/1550/#showFailuresLink

### PR DESCRIPTION
**Pull request by murgatroid99**
This fixes #10879.

The main fixes are the change in `tcp_server_uv.c` to wait to accept incoming connections until the server starts, and the change in `iomgr_uv.c` to disable threading in the executor.

I also added a check to several libuv-specific iomgr functions to assert that they are running on the same thread that initialized the library, to try to avoid this in the future, or at least make it easier to debug.

There are a couple of other fixes to minor memory leaks and uses of uninitialized variables that were discovered while debugging this issue, and I disabled one test from running in the libuv tests because it starts a thread.

### CODE REVIEW

File: src/core/lib/iomgr/iomgr_uv.c [hunk 1]
```diff
@@ -21,12 +21,20 @@
 #ifdef GRPC_UV
 
 #include "src/core/lib/debug/trace.h"
+#include "src/core/lib/iomgr/executor.h"
+#include "src/core/lib/iomgr/iomgr_uv.h"
 #include "src/core/lib/iomgr/pollset_uv.h"
 #include "src/core/lib/iomgr/tcp_uv.h"
 
+gpr_thd_id grpc_init_thread;
```
Comment (nicolasnoble) [inline src/core/lib/iomgr/iomgr_uv.c]: prefix global variables with a `g_`.

File: src/core/lib/iomgr/iomgr_uv.h [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+#include "src/core/lib/iomgr/iomgr_internal.h"
+
+#include <grpc/support/thd.h>
+
+/* The thread ID of the thread on which grpc was initialized. Used to verify
+ * that all calls into libuv are made on that same thread */
+extern gpr_thd_id grpc_init_thread;
+
+#ifdef GRPC_UV_THREAD_CHECK
+#define GRPC_ASSERT_SAME_THREAD() \
```
Comment (nicolasnoble) [inline src/core/lib/iomgr/iomgr_uv.h]: maybe insert `UV_` somewhere in there ?

File: src/core/lib/iomgr/tcp_server_uv.c [hunk 3]
```diff
@@ -142,10 +147,12 @@ static void handle_close_callback(uv_handle_t *handle) {
 }
 
 static void close_listener(grpc_tcp_listener *sp) {
+  grpc_exec_ctx exec_ctx = GRPC_EXEC_CTX_INIT;
   if (!sp->closed) {
     sp->closed = true;
     uv_close((uv_handle_t *)sp->handle, handle_close_callback);
   }
+  grpc_exec_ctx_finish(&exec_ctx);
```
Comment (nicolasnoble) [inline src/core/lib/iomgr/tcp_server_uv.c]: Does `uv_close` call its callback immediately potentially so you had to add a call to `grpc_exec_ctx_finish` here ?

File: src/core/lib/iomgr/tcp_server_uv.c [hunk 4]
```diff
@@ -207,35 +246,17 @@ static void on_connect(uv_stream_t *server, int status) {
...
-  client = gpr_malloc(sizeof(uv_tcp_t));
-  uv_tcp_init(uv_default_loop(), client);
-  // UV documentation says this is guaranteed to succeed
-  uv_accept((uv_stream_t *)server, (uv_stream_t *)client);
-  // If the server has not been started, we discard incoming connections
-  if (sp->server->on_accept_cb == NULL) {
-    uv_close((uv_handle_t *)client, accepted_connection_close_cb);
+  GPR_ASSERT(!sp->has_pending_connection);
+
+  gpr_log(GPR_DEBUG, "SERVER_CONNECT: %p incoming connection", sp->server);
```
Comment (nicolasnoble) [inline src/core/lib/iomgr/tcp_server_uv.c]: Maybe add a `grpc_tcp_trace` around that one ?

File: src/core/lib/iomgr/tcp_server_uv.c (hunk 3 above)
Comment (murgatroid99) [inline src/core/lib/iomgr/tcp_server_uv.c]: I don't know why I added that. I'm pretty sure it doesn't do anything. That `exec_ctx` is never used.

**Comment by murgatroid99** [comment]
Asan and Msan actually passed, but reported failure. The Windows failure is #11860. Based on a local run, the sanity failure looks related to OWNERS, but I didn't touch that in this PR. And the interop tests almost all passed, but one timed out.

**Pull request merged**